### DISCUSSION TITLE
supress "should implement method" errors

### DISCUSSION THREAD

**Discussion by kor3k** [symfony/symfony#48016] [Q&A]
hello,

i have denoted an optional method in interface via phpdoc `@method`

```
/**
 * @method string getHashSalt()
 */
interface HashableInterface
```

it is supposed to be **optional**, ie class may or may not implement it.

but i have these annoying errors everywhere

```
User Deprecated: Class "App\Entity\User" should implement method "App\Model\HashableInterface::getHashSalt(): string".
```

is there a way to suppress them? and i mean just this context, regarding "should implement method", not "User Deprecated" in general.

**Comment by ghost** [comment]
There is no such thing as **optional** method. `@method` annotation is the same as real method but used for magic methods or for BC layers.

**Reply by kor3k** [reply]
well, i both get and don't get your point :) 
i understand this `@method` is usually used together wih call magic. but, php is a dynamic language, with things like `method_exists()`, so there really should be something to denote an optional method. and because there is nothing else than `@method` in phpdoc, i have to use this annotation. that's for the rationale of my question.

but anyway, even if i "abuse" this `@method` annot, is there a way to suppress those errors? because i have not just one of those "optional" methods, and these errors pollute logs and shell output.

**Reply by ghost** [reply]
I've never seen an optional method in my whole life.
Why can't you just return null from `getHashSalt` what would denote there is no hash salt and classes that right  now do not require this method would return null?
I think your code design is poor. Why would you implement this interface and sometimes implement this method and sometimes not? Maybe you need another interface like `HashableWithSaltInterface` (couldn't figure out better name) that would extends `HashableInterface` and implement this one in classes that need this method?

**Reply by kor3k** [reply]
> Why can't you just return null from getHashSalt 

first, it's not just this specific method/interface, there are more of them for different things.
second, i use this with `method_exists(string $class)`, ie without instance.

> Maybe you need another interface like HashableWithSaltInterface (couldn't figure out better name) that would extends HashableInterface and implement this one in classes that need this method?

i actually had it this way. but having "duplicate" interfaces for just one method seemed untidy, when it can be replaced with just one `method_exists` check. 

> I think your code design is poor. Why would you implement this interface and sometimes implement this method and sometimes not?

it works well. the only thing are those polluting errors. why? given this specific "hash" example - i check if it has this method. if yes, i have to get the instance/entity (it generates it's own salt). if not, i use the classname as salt.

**Comment by ghost** [comment]
>  i use this with method_exists(string $class), ie without instance.

I can see you optimize you code for writing to save some key strokes instead for reading. If you really want to work without an instance as long as possible then I see 2 ways:
1.  add another method to the interface that is static, e.g. `hasSalt()` and you can call it without an instance. If this returns `true` then you fetch instance to call `getHashSalt`, if `false` then use class name.
2. Add another interface as I said before and use function `is_subclass_of()`. This does not cost you anything and you get a better desing.
